### PR TITLE
Add emoji reactions to messages

### PR DESCRIPTION
**Pull request by mrkvon**
- ~Include library for selecting emoji `emoji-mart-vue`~
- [x] See who reacted (in tooltip)
- [x] see whether i reacted or not (color the emoji background if i reacted)
- [x] write tests
- [ ] websocket

### CODE REVIEW

**Comment by mrkvon** [comment]
Now this PR + yunity/karrot-backend#507 probably needs to be reviewed in order to move further. Emoji reactions work, but can be improved. My knowledge is not yet large enough to improve the code with reasonably low level of frustration.

@tiltec @nicksellen would you have some suggestions?

File: src/store/modules/conversations.js [hunk 1]
```diff
@@ -97,6 +105,26 @@ export default {
         await conversationsAPI.toggleEmailNotifications(conversationId, value)
         commit('updateEmailNotifications', { conversationId, value })
       },
+
+      /**
+       * Add reaction to a message.
+       */
+      async editReaction ({ state, commit }, { action, name, messageId, userId }) {
```
Comment (tiltec) [inline src/store/modules/conversations.js]: Maybe you don't need to pass the user id? Rather always assume that the current user did it  -> rootGetters['auth/userId']

Also, I wouldn't pass the action, rather make it a `toggleReaction` - the vuex module has all information to find out if the reaction already exists and which endpoint to call.

Generally, I'd keep the interface between vuex and components as small as possible and strip logic from components. Vuex modules are easier to test and easier to reuse.

**Comment by mrkvon** [comment]
@tiltec the issues from review should be (hopefully) addressed.

I would still like to try replacing the custom emoji selection with `Picker` from `emoji-mart-vue` library. Stubbing it could make the tests pass. It could be done here or in another PR.

**Comment by mrkvon** [comment]
The database of emoji names in the backend often doesn't match the `emoji-mart-vue` names. It could be a headache to make them compatible. Not sure how to proceed beyond the limited diy list of emojis which is there now.

**Comment by tiltec** [comment]
Does `emoji-mart-vue` have more emojis than the backend list or are they just different names? Are the aliases not enough?

**Comment by mrkvon** [comment]
> Does `emoji-mart-vue` have more emojis than the backend list or are they just different names? Are the aliases not enough?

It's probably both. On the backend we always save the emoji name in its default form according to the backend emoji database.
Sometimes this default form is not recognized by `emoji-mart-vue`, so it's saved but the emoji is not displayed. (We use emoji-mart-vue for displaying added emojis, too.)
Sometimes the name provided by emoji-mart-vue Picker isn't recognized by the backend at all (we look it up both in the default names and aliases).

File: src/components/Conversation/ConversationReactions.vue [hunk 2]
```diff
@@ -0,0 +1,147 @@
+<template>
+  <!-- emoji reactions to messages -->
+  <div class="message-reactions">
+    <!-- reactions -->
+    <q-btn
+      class="reaction-button emoji-button"
+      @click.native="toggleReaction(reaction.name)"
```
[the remark this answers is not in the document]
Comment (mrkvon) [reply src/components/Conversation/ConversationReactions.vue]: I've probably seen it somewhere. Removed.

Sorry for the long reaction times. I haven't been very computer-productive last days.

[the remark this answers is not in the document]
Comment (tiltec) [reply src/components/Conversation/ConversationReactions.vue]: No worries, it was also quite busy here :)

**Comment by tiltec** [comment]
A quick design & usability review:

- [ ] it could be clearer which message the reaction belong to
- [ ] the dark-blue background color bites with yellow of the emojis if the user reacted themselves
- [ ] when you click on the "+" button, it shouldn't show emojis that you already reacted with. alternatively, show them but don't remove the reaction when the user clicks on it
- [ ] the "+" buttons could take up less space. maybe only show them when hovering over the message (like slack). the mobile view could get a popover to add reactions when clicking on the message, then the "+" button could go away

image
image
image

Are you motivated to play with the design a bit @mrkvon ? If not, it's no problem at all. You already worked a lot on this, so I'm happy to jump in here. Or maybe @D0nPiano wants to give it a try :)

**Comment by mrkvon** [comment]
@tiltec I'm a bit challenged when it comes to design/css tinkering/looks.
If there is anybody who likes that part, would be great if they take it over. Otherwise I can try. I'm still happy to fix the functional part.

>  when you click on the "+" button, it shouldn't show emojis that you already reacted with. alternatively, show them but don't remove the reaction when the user clicks on it

This is how slack and github emojis work. (remove existent emoji on click in menu)

Also, the `emoji-mart-vue` library is currently used for displaying emojis. It may be worth removing that dependency if we're not going to use its `Picker` element.
However, it was challenging for me to style the buttons when the `Markdown` element was used for displaying emojis. Not sure which way to go.

**Comment by tiltec** [comment]
> This is how slack and github emojis work. (remove existent emoji on click in menu)

Indeed. It seems bad UX to me though.

> Also, the emoji-mart-vue library is currently used for displaying emojis. It may be worth removing that dependency if we're not going to use its Picker element.

I'd still like to use some full-blown emoji picker at some point. The biggest problem with `emoji-mart-vue` was non-matching backend name validation, right? I wonder if it would be sensible to remove the name validation. What problems could happen if we do?

> I'm a bit challenged when it comes to design/css tinkering/looks.

I feel you, I'm usually frustrated about my design skills. Sometimes it works quite well though, so I'm happy to give it a try!

**Comment by mrkvon** [comment]
> Indeed. It seems bad UX to me though.

Do you want it changed?

> I wonder if it would be sensible to remove the name validation. What problems could happen if we do?

Validation in this place serves two purposes:
1. Don't allow custom emoji names which can't be displayed (i.e. `foo_bar`).
2. If an emoji can have aliases (`thumbsup`, `+1`), always save the default name.

The main problems that can occur:
1. We'll have buttons without emojis in the list of given reactions.
2. We'll be able to have two thumbsups (even given by one person) in the list next to each other (if both `+1` and `thumbsup` reactions were given)

Both issues appear only if api is accessed outside of app, or if the Picker changes.

**Comment by tiltec** [comment]
> Do you want it changed?

I feel there's a better solution, but I may have to experiment with it before it's clearer to me. Feel free to go forward with changing it if you're motivated :)


Thanks for describing the problems when leaving validation away, I agree that it makes sense to have it if we would recommend direct API access or have multiple clients connected. As I see it, we will have only one client for a long time coming, so maybe we go the more convenient route of leaving backend validation away.
If problems occur in future, we can always implement a more suitable backend validation and run a data migration to fix up existing data.

**Comment by tiltec** [comment]
I had a look again and would now get rid of `emoji-mart-vue` because of problems with backend validation and the increase in bundle size of 93kB gzipped: https://bundlephobia.com/result?p=emoji-mart-vue@2.4.5

Instead, I would preselect some emojis like it is right now (possibly rendered by the Markdown component) and for more emojis, add a autocomplete-based picker like Telegram:

image
(Maybe we could also reuise this for markdown-input at some point :) )

I'll work on this.

**Comment by tiltec** [comment]
@mrkvon thanks!
I was playing a bit with `twemoji` and `markdown-it-emoji` and it seems we don't need the `markdown` component for it. I'll push a commit soon :)

**Comment by tiltec** [comment]
I tried out the Slack style, what do you think about it?

image

A bit more margin between the messages could be good. It's not completely finished, will work more on it tomorrow.

File: src/storyshots.spec.js [hunk 3]
```diff
@@ -5,6 +5,7 @@
  * Therefore, we mimick the Storybook API to get the components and then run the snapshot tests
 */
 let mockStories = []
+const consoleErrorSpy = jest.spyOn(global.console, 'error')
```
Comment (tiltec) [inline src/storyshots.spec.js]: This should catch `console.log` messages in storyshot tests and let the test fail if one occurs.

**Comment by mrkvon** [comment]
I think it is very nice now. I like both the matching highlight style and the appearing of the `+` button both in the corner and next to emojis.

Thank you @tiltec for this work, for which my experience and patience may have been too limited. Not only the visual changes, but also (and importantly) the refactor! :+1:

**Comment by mrkvon** [comment]
:tada:

**Pull request merged**